vault/resource_kubernetes_auth_backend_config.go

```diff
@@ -0,0 +1,218 @@
+		}
+		data["pem_keys"] = strings.Join(pemKeys, ",")
+	}
+	data["kubernetes_host"] = d.Get("kubernetes_host").(string)
+
+	_, err := client.Logical().Write(path, data)
+	if err != nil {
+		return fmt.Errorf("error writing Kubernetes auth backend config %q: %s", path, err)
+	}
+
+	d.SetId(path)
+	// NOTE: Since reading the auth/<backend>/config does
+	// not return the `token_reviewer_jwt`,
+	// set it from data after successfully storing it in Vault.
+	d.Set("token_reviewer_jwt", data["token_reviewer_jwt"])
+	log.Printf("[DEBUG] Wrote Kubernetes auth backend config %q", path)
+
+	return kubernetesAuthBackendConfigRead(d, meta)
+}
+
+func kubernetesAuthBackendConfigBackendFromPath(path string) (string, error) {
+	if !kubernetesAuthBackendConfigFromPathRegex.MatchString(path) {
+		return "", fmt.Errorf("no backend found")
+	}
+	res := kubernetesAuthBackendConfigFromPathRegex.FindStringSubmatch(path)
+	if len(res) != 2 {
+		return "", fmt.Errorf("unexpected number of matches (%d) for backend", len(res))
+	}
+	return res[1], nil
+}
+
+func kubernetesAuthBackendConfigRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+	path := d.Id()
+
+	backend, err := kubernetesAuthBackendConfigBackendFromPath(path)
+	if err != nil {
+		return fmt.Errorf("invalid path %q for Kubernetes auth backend config: %s", path, err)
+	}
+
+	log.Printf("[DEBUG] Reading Kubernetes auth backend config %q", path)
+	resp, err := client.Logical().Read(path)
+	if err != nil {
+		return fmt.Errorf("error reading Kubernetes auth backend config %q: %s", path, err)
+	}
+	log.Printf("[DEBUG] Read Kubernetes auth backend config %q", path)
+	if resp == nil {
+		log.Printf("[WARN] Kubernetes auth backend config %q not found, removing from state", path)
+		d.SetId("")
+		return nil
+	}
+
+	d.Set("backend", backend)
+	d.Set("kubernetes_host", resp.Data["kubernetes_host"])
+	d.Set("kubernetes_ca_cert", resp.Data["kubernetes_ca_cert"])
+
+	iPemKeys := resp.Data["pem_keys"].([]interface{})
+	pemKeys := make([]string, 0, len(iPemKeys))
+
+	for _, iPemKey := range iPemKeys {
+		pemKeys = append(pemKeys, iPemKey.(string))
+	}
+
+	d.Set("pem_keys", pemKeys)
+
+	return nil
+}
+
+func kubernetesAuthBackendConfigUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+	path := d.Id()
+
+	log.Printf("[DEBUG] Updating Kubernetes auth backend config %q", path)
+
+	data := map[string]interface{}{}
+
+	if v, ok := d.GetOk("kubernetes_ca_cert"); ok {
+		data["kubernetes_ca_cert"] = v.(string)
+	}
+
+	if v, ok := d.GetOk("token_reviewer_jwt"); ok {
+		data["token_reviewer_jwt"] = v.(string)
+	}
+
+	if v, ok := d.GetOkExists("pem_keys"); ok {
+		var pemKeys []string
+		for _, pemKey := range v.([]interface{}) {
+			pemKeys = append(pemKeys, pemKey.(string))
+		}
+		data["pem_keys"] = strings.Join(pemKeys, ",")
+	}
+	data["kubernetes_host"] = d.Get("kubernetes_host").(string)
+
+	_, err := client.Logical().Write(path, data)
+
+	d.SetId(path)
```

Comment:
Would it be possible to move this up above the `Write` line so it's not between the err creation and the error check?

Reply:
Good note. Actually it's a mistake on my end. 

It must be after the `Write`. Reasoning behind this is to make sure that `SetId` is called only afteer the value has been successfully written in Vault.